inductor: using binary folding path to do conv+bn folding

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105651
* __->__ #105650
* #105048

This path will use binary folding to do conv+bn folding to avoid using ```make_fx``` which meets tracing errors in some model dynamic shape path.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

XiaobingSuper committed:
inductor: using binary folding path to do conv+bn folding

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: using binary folding path to do conv+bn folding"


This path will use binary folding to do conv+bn folding to avoid using ```make_fx``` which meets tracing errors in some model dynamic shape path.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: using binary folding path to do conv+bn folding"


This path will use binary folding to do conv+bn folding to avoid using ```make_fx``` which meets tracing errors in some model dynamic shape path.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison commented on torch/_inductor/fx_passes/freezing_patterns.py:
```diff
@@ -22,14 +24,23 @@
 
 
-def freezing_passes(gm: torch.fx.GraphModule):
+def freezing_passes(gm: torch.fx.GraphModule, aot_example_inputs):
```
can we special case this and just run the folding pass for a number of iterations first ? feels unnecessary to run each pass 4 times. 

We should also be able to short-circuit if the binary folding path didn't find any additional matches

XiaobingSuper replied:
Yes,  separate binary folding pass and short-circuit if the binary folding path finds no additional matches.

eellison reviewed (commented):
cool!

XiaobingSuper committed:
Update on "inductor: using binary folding path to do conv+bn folding"


This path will use binary folding to do conv+bn folding to avoid using ```make_fx``` which meets tracing errors in some model dynamic shape path.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: using binary folding path to do conv+bn folding"


This path will use binary folding to do conv+bn folding to avoid using ```make_fx``` which meets tracing errors in some model dynamic shape path.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

XiaobingSuper commented on test/inductor/test_inductor_freezing.py:
```diff
@@ -292,7 +292,10 @@ def foo(mod, x):
+        self.assertNotIn(
+            "aten._native_batch_norm_legit_no_training(",
+            code[0],
+        )
```
Change the code because the generated code has ```# Source Nodes: [l__mod___bn], Original ATen: [aten._native_batch_norm_legit_no_training]```.

XiaobingSuper committed:
Update on "inductor: using binary folding path to do conv+bn folding"


This path will use binary folding to do conv+bn folding to avoid using ```make_fx``` which meets tracing errors in some model dynamic shape path.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: using binary folding path to do conv+bn folding"


This path will use binary folding to do conv+bn folding to avoid using ```make_fx``` which meets tracing errors in some model dynamic shape path.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison commented on torch/_inductor/fx_passes/freezing_patterns.py:
```diff
@@ -24,15 +27,31 @@
+    # We need a few rounds of binary folding to get rid of all the
+    # unnecessary nodes, but may need a good method to chose the rounds number.
+    # works like: conv+binary+binary.
+    binary_folding = 0
```
im not 100% sure the counters get reset so i would initialize `binary_folding = counters["inductor"]["binary_folding"]`

XiaobingSuper replied:
changed.

eellison commented on torch/_inductor/fx_passes/freezing_patterns.py:
```diff
@@ -24,15 +27,31 @@
+    for _ in range(4):
+        constant_fold(gm)
+        # Make sure meta['val'] is properly set for all nodes
+        fake_tensor_prop(gm, aot_example_inputs, True)
```
TODO: remove the need to run fake_tensor_prop on the whole model,

XiaobingSuper replied:
added.

eellison reviewed (approved):
🚀

XiaobingSuper committed:
Update on "inductor: using binary folding path to do conv+bn folding"


This path will use binary folding to do conv+bn folding to avoid using ```make_fx``` which meets tracing errors in some model dynamic shape path.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge

eellison commented on torch/_inductor/fx_passes/freezing_patterns.py:
```diff
@@ -24,15 +27,32 @@
+        # TODO: remove the need to run fake_tensor_prop on the whole model.
+        if counters["inductor"]["binary_folding"] == binary_folding:
+            break
+        binary_folding += counters["inductor"]["binary_folding"]
```
From looking at `res2next50`, i think we still need to add support for no bias
